st2common/st2common/util/crypto.py

```diff
@@ -13,40 +13,291 @@
+                      hmac_key_size=key_size, mode='CBC', size=key_size)
+
+    def __json__(self):
+        """
+        Return JSON representation of this key which is fully compatible with keyczar JSON key
+        file format.
+
+        :rtype: ``str``
+        """
+        data = {
+            'hmacKey': {
+                'hmacKeyString': self.hmac_key_string,
+                'size': self.hmac_key_size
+            },
+            'aesKeyString': self.aes_key_string,
+            'mode': self.mode.upper(),
+            'size': int(self.size)
+        }
+        return json.dumps(data)
+
+    def __repr__(self):
+        return ('<AESKey hmac_key_size=%s,mode=%s,size=%s>' % (self.hmac_key_size, self.mode,
+                                                               self.size))
+
+
+def read_crypto_key(key_path):
     """
+    Read crypto key from keyczar JSON key file format and return parsed AESKey object.
+
+    :param key_path: Absolute path to file containing crypto key in Keyczar JSON format.
+    :type key_path: ``str``
 
-    # Late import to avoid very expensive in-direct import (~1 second) when this function
-    # is not called / used
-    from keyczar.keys import AesKey
-    key_type = key_type or AesKey
+    :rtype: :class:`AESKey`
+    """
+    with open(key_path, 'r') as fp:
+        content = fp.read()
 
-    with open(key_path) as key_file:
-        key = key_type.Read(key_file.read())
-        return key
+    content = json.loads(content)
+
+    try:
+        aes_key = AESKey(aes_key_string=content['aesKeyString'],
+                         hmac_key_string=content['hmacKey']['hmacKeyString'],
+                         hmac_key_size=content['hmacKey']['size'],
+                         mode=content['mode'].upper(),
+                         size=content['size'])
+    except KeyError as e:
+        msg = 'Invalid or malformed key file "%s": %s' % (key_path, str(e))
+        raise KeyError(msg)
+
+    return aes_key
 
 
 def symmetric_encrypt(encrypt_key, plaintext):
+    return cryptography_symmetric_encrypt(encrypt_key=encrypt_key, plaintext=plaintext)
+
+
+def symmetric_decrypt(decrypt_key, ciphertext):
+    return cryptography_symmetric_decrypt(decrypt_key=decrypt_key, ciphertext=ciphertext)
+
+
+def cryptography_symmetric_encrypt(encrypt_key, plaintext):
+    """
+    Encrypt the provided plaintext using AES encryption.
+
+    NOTE 1: This function return a string which is fully compatible with Keyczar.Encrypt() method.
+
+    NOTE 2: This function is loosely based on keyczar AESKey.Encrypt() (Apache 2.0 license).
+
+    The final encrypted string value consists of:
+
+    [message bytes][HMAC signature bytes for the message] where message consists of
+    [keyczar header plaintext][IV bytes][ciphertext bytes]
+
+    NOTE: Header itself is unused, but it's added so the format is compatible with keyczar format.
+
+    """
+    assert isinstance(encrypt_key, AESKey), 'encrypt_key needs to be AESKey class instance'
```

Comment:
These asserts look okay, just be aware that running Python with optimizations turned on bypasses all assertions made with the `assert` keyword. This has caused security issues in the past ([example: PySAML](https://github.com/IdentityPython/pysaml2/issues/451)).

[Assertions are meant to be used as debugging aids, not as a mechanism for error handling.](https://www.quora.com/How-do-I-use-Python-asserts-even-when-optimization-mode-is-used/answer/Dan-Bader-1)

It sucks, but you may want to use plain `if` statements for all of your assertions:

```python
if not isinstance(plaintext, (six.text_type, six.string_types, six.binary_types)):
    raise CustomErrorType(...)  # Not an AssertionError
```

More help: https://dbader.org/blog/python-assert-tutorial

Reply:
Yeah, I used them for what they are intended - as debugging / developer aid aka to make sure programmer passes in a correct type (stuff which should get caught during tests and not during run time for which I would use exceptions).

For dynamic run time stuff I always use exceptions and not asserts.